### PR TITLE
WIP: Start refactoring of esil engine

### PR DESCRIPTION
**Pull request by condret**
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

### CODE REVIEW

File: libr/include/r_esil.h [hunk 1]
```diff
@@ -148,6 +148,29 @@ typedef struct r_esil_callbacks_t {
 	bool (*reg_write)(ESIL *esil, const char *name, ut64 val);
 } REsilCallbacks;
 
+typedef bool (*REsilMemSwitch)(void *mem, ut32 idx);
+typedef bool (*REsilMemRead)(void *mem, ut64 addr, ut8 *buf, int len);
+typedef bool (*REsilMemWrite)(void *mem, ut64 addr, const ut8 *buf, int len);
+
+typedef struct r_esil_memory_interface_t {
+	void *mem;
```
Comment (trufae) [inline libr/include/r_esil.h:156]: userdata?

File: libr/include/r_esil.h [hunk 2]
```diff
@@ -148,6 +148,29 @@ typedef struct r_esil_callbacks_t {
...
+	REsilMemSwitch mem_switch;
```
[the remark this answers is not in the document]
Comment (condret) [reply libr/include/r_esil.h:157]: yes, but it's optional, so you could also use something that doesn't provide banking

File: libr/include/r_esil.h (hunk 1 above)
Comment (condret) [reply libr/include/r_esil.h:156]: yes, within r2 this would be RIO *

**Comment by trufae** [comment]
build fails because you cant use io apis from esil library, because it hasnt been built yet. you can use RIOBind instead or try to mess with the library dependency order in library/libs.mk

File: libr/include/r_esil.h (hunk 1 above)
Comment (trufae) [reply libr/include/r_esil.h:156]: Better user userdata instead of mem for cosistency

Comment (condret) [reply libr/include/r_esil.h:136]: will change that later

**Pull request merged**